javascript/javascript-maps.md

```diff
@@ -0,0 +1,110 @@
+# Javascript Maps
+
+## Projected Time
+
+30-45 minutes
+
+## Prerequisites
+
+- [Javascript 6 Lesson](https://github.com/Techtonica/curriculum/blob/master/javascript/javascript-6-object-literals.md)
+
+## Motivation
+
+Map object provided by ES6. Map is a collection of elements where each element is stored as a Key, value pair. Map object can hold both objects and primitive values as either key or value. You can use objects of all kinds as map keys. The system will not automatically convert the keys to strings as it does for object literals. This opens up a wide range of opportunities to do interesting things with maps.
+
+## Objectives
+
+**Participants will be able to:**
+
+- Learn maps.
+- Learn sets.
+- Know the difference between maps and objects.
+- Create and use maps & sets.
+
+## Specific Things To Teach
+
+- Maps.
+   - Syntax.
+   - Methods or operations.
+   - Example.
+   - When to use maps or objects!
+- Sets.
+   - Syntax.
+   - Example.
+- Maps and objects compared.
+
+## Materials
+
+- [JavaScript ES6 - Maps (3 min video)](https://www.youtube.com/watch?v=QjYk58e-8v4) - This video walks through ES6 Map's. How they are similar and different to Objects and how to use them!
+
+- [Techtonica slides](https://docs.google.com/presentation/d/1fL2vgcI4uZPayAa-qCa34Rnuo0I3rfZIf7Lm7nnhNcU/edit?usp=sharing) - Slide on javascript maps and sets.
+
+## Lesson
+
+**Maps** are a new object type that allow to store collections of key-value pairs. Unlike with objects, map keys can be of any type, even objects or functions. With maps we can iterate in the order in which the values were added, unlike objects where there’s no guarantee about the order.
+
+* Syntax:<br>
+  	```javascript
+	new Map([iterable]);
+	```
+
+* The main methods are:
+	- **new Map()** – creates the map.
+	- **map.set(key, value)** – stores the value by the key.
+	- **map.get(key)** – returns the value by the key, undefined if key doesn’t exist in map.
+	- **map.has(key)** – returns true if the key exists, false otherwise.
+	- **map.delete(key)** – removes the value by the key.
+	- **map.clear()** – clears the map.
+	- **map.size** – returns the current element count.
+
+* Example:<br>
+	```javascript
+	let things = new Map();
+	things.set('🚗', 'Car');
+	things.set('🏠', 'House');
+	things.size; // 2
+	things.clear();
+	things.size; // 0
+	```
+
+* The differences from a regular Object:
+
+	- Keys can be objects too in maps.
```

Comment:
Please make these into complete sentences and not fragments. Beginners won’t be able to interpret this.

Reply:
They likely don’t know that objects can only have numbers or strings as keys; please tell them here.

Reply:
It should be just like your when to use map? Section